Download picture from gallery

rawbertp opened the pull request:
Allow download of pictures from gallery.

![Screenshot 2019-11-06 at 08 21 54](https://user-images.githubusercontent.com/12150805/68277094-c3d2dc00-006e-11ea-8e31-9cd08d777fbd.png)

Should fix #7 .

andi34 commented:
Could we make this an option in admin panel?

rawbertp commented:
@sualko Please review the `img` refactoring - it works but I am not sure if this is the correct approach!

@andi34 - sure why not ;)

andi34 commented:
You'll take care of it? Or should I look at it tomorrow?

rawbertp commented:
I'm on it. However, I was just thinking that it wouldn't make sense to allow QR but no download and vice versa. I would combine it: if DL is enabled, both will be shown QR and DL button. OK with that?

rawbertp commented:
I have now implemented it as separate option. Still, I don't think this makes sense, but it can be easily changed once there is an agreement. ;)

andi34 commented:
I'd make it an own option independent of QR, gives people all possibilities to fit personal needs.
Maybe someone likes none of it, for me only the DL button would be used (no one i know really uses QR codes).

andi34 commented:
Just tested, working fine :) thanks a lot!!!

andi34 commented:
Only one thing we could change is to hide the button if we're accessing the page from our Pi (hide on access from IP of the Pi,  localhost and 127.0.0.1).

andi34 commented on resources/js/photoinit.js:
```diff
@@ -178,12 +175,14 @@ function initPhotoSwipeFromDOM (gallerySelector) {
     });
 
+    $('.pswp__button--download').on('click touchstart', function (e) {
+        e.preventDefault();
```
we might also need `e.stopPropagation();` , see https://github.com/andreknieriem/photobooth/issues/185

rawbertp replied:
Done.

andi34 commented:
maybe you could cherry-pick and push this one here too https://github.com/andi34/photobooth/commit/366ea6661cd83c7087b810508974ddccf972e122

rawbertp commented:
@andi34 I added a check for `SERVER_ADDR != REMOTE_ADDR` and added the `stopPropagation` call. The other change you mentioned will get merged anyway, so I didn't cherry-pick it.

sualko commented on resources/js/photoinit.js:
```diff
@@ -3,6 +3,7 @@
 function initPhotoSwipeFromDOM (gallerySelector) {
 
     let gallery;
+    let img;
```
I don't like global variables, because they make the code less readable.

rawbertp replied:
I don't like the code duplication but I get your point and will revert these changes. It should be improved IMHO but has nothing to do with that PR.

sualko commented on resources/js/photoinit.js:
```diff
@@ -132,6 +133,8 @@ function initPhotoSwipeFromDOM (gallerySelector) {
 
         gallery.init();
+
+        img = gallery.currItem.src.split('/').pop();
```
The further use of `img` is now depending on `parseThumbnailElements` which is a bad idea in my opinion.

sualko commented on resources/js/photoinit.js:
```diff
@@ -178,12 +175,15 @@ function initPhotoSwipeFromDOM (gallerySelector) {
+    $('.pswp__button--download').on('click touchstart', function (e) {
+        e.preventDefault();
+        e.stopPropagation();	
+        location = 'api/download.php?image=' + img;
```
`window.location` would be better than `location` (because it makes clear that `location` is no local variable), but the [html5 download](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download) attribute would be even better. If you create a new `a` element at this position with the corresponding attributes a emulated click event on that element should work.

rawbertp replied:
If I get you right, you suggest to replace `<button class="pswp__button pswp__button--download" ...` in `pwswp.template.php` (i.e. the current approach that applies to all the other buttons there) with an `<a href="" download="<js-injected-url>" ...` ?

sualko replied:
No, I meant something like this (untested):

```
const linkElement = $('<a>');
linkElement.attr({
   href: 'URL_TO_IMAGE',
   download: 'FILENAME',
});
linkElement.click();
```

This should ensure that the browser shows the download prompt instead of open the image.

andi34 replied:
Also check if download attribute is supported by the browser?
```
function doesAnchorSupportDownload() {
    return "download" in document.createElement("a")
}
```
Currently I am checking it this way
```
        if (doesAnchorSupportDownload()) {
            location = 'api/download.php?image=' + img;
        } else {
            alert('Unsupported on your Browser');
        }
```

rawbertp replied:
Sorry, I’m afraid I don’t get it. „On click“ the element gets manipulated and then .click() gets called? What’s the benefit?

I guess I will leave this PR open and leave it up to you (or s.o. else with web developer skills) to properly implement this feature. 😌

rawbertp replied:
@sualko Out of curiosity I still tried to fix this according to your suggestions, see: https://github.com/andreknieriem/photobooth/pull/183/commits/7d986471fd54b9f9b1e5f1f0f49c0102f64a651c . It works but might be complete bullsh... Looking forward to your feedback. ;-)

andi34 commented:
Also check if download attribute is supported by the browser?
```
function doesAnchorSupportDownload() {
    return "download" in document.createElement("a")
}
```
Currently I am checking it this way
```
        if (doesAnchorSupportDownload()) {
            location = 'api/download.php?image=' + img;
        } else {
            alert('Unsupported on your Browser');
        }
```

sualko commented:
That was smart. Even better than my solution. I made only a couple of minor changes, otherwise a great pr. Thanks a lot.

@andi34 there is no need to check for support. If the browser doesn't know the attribute it get's ignored.

andi34 commented:
@sualko but wouldn't it be better to tell the people instead silent not working?

rawbertp commented:
@sualko Thanks, but that was a bit too quick now, I guess... You merged just the first but not the latest commit. I suggest to squash these on my branch and then merge the PR?

andi34 commented:
@rawbertp He did already ;) https://github.com/andreknieriem/photobooth/commit/c4ace13c6913fa6d414d7b92d7dca124509015e4

rawbertp commented:
Ah ok i see.. Okay but there might be one minor issue left with the button. I haven't looked into it yet as I first wanted to wait if this gets accepted at all. IIRC the button was not correctly aligned with the others - I will have to have a look...

andi34 commented:
@rawbertp haven't checked yet, but he also added the line hight here:
https://github.com/andreknieriem/photobooth/commit/c4ace13c6913fa6d414d7b92d7dca124509015e4#diff-28f3b1da8414261c1172ba056d85e851R13

rawbertp commented:
Alright, looks good (also on the iPad). :)

sualko commented:
> He did already

Yes, he did :joy:

>  but wouldn't it be better to tell the people instead silent not working?

That's no problem. The browser will still show the picture.
